internal/engine/k8swatch/pod_watch.go

```diff
@@ -152,19 +157,46 @@ func (w *PodWatcher) removeWatch(toRemove PodWatch) {
 	}
 }
 
+// We inject a label to all pod template specs so that we can tell if it came from a pod template we've deployed
+// since startup.
+// This is to solve the problem where:
+// 1. start Tilt, which applies a deployment and creates a pod
+// 2. kill Tilt, change some code that goes into the pod's image
+// 3. start Tilt, which reapplies that same deployment
+// At this point, we want a new pod to be created, and we don't care about the old pod. However, since
+// changing a deployment's PodTemplateSpec does not cause the deployment to get a new UID, the deployment
+// applied in (3) has the same UID as the deployment applied in (1), so when we get an event for the
+// undesired pod from (1), we trace it back to the deployment we applied in (3), and emit an event.
+// To deal with this, we inject a hash of the pod template spec into the pod template itself, and only
+// emit events from pods with known pod template spec hashes.
+func hasOKPodTemplateSpecHash(pod *v1.Pod, set store.PodTemplateSpecHashSet) bool {
+	// if it doesn't have a label, just let it through - maybe it's from a CRD w/ no pod template spec
+	hash, ok := pod.Labels[k8s.TiltPodTemplateHashLabel]
+	if !ok {
+		return true
+	}
+
+	return set.Contains(k8s.PodTemplateSpecHash(hash))
+}
+
 // When new UIDs are deployed, go through all our known pods and dispatch
 // new actions. This handles the case where we get the Pod change event
 // before the deploy id shows up in the manifest, which is way more common than
 // you would think.
-func (w *PodWatcher) setupNewUIDs(ctx context.Context, st store.RStore, newUIDs map[types.UID]model.ManifestName) {
+func (w *PodWatcher) setupNewUIDs(ctx context.Context, st store.RStore, newUIDs map[types.UID]model.ManifestName, podTemplateSpecHashes map[model.ManifestName]store.PodTemplateSpecHashSet) {
 	w.mu.Lock()
 	defer w.mu.Unlock()
 
 	for uid, mn := range newUIDs {
 		w.knownDeployedUIDs[uid] = mn
 
 		pod, ok := w.knownPods[uid]
+
 		if ok {
+			set, ok := podTemplateSpecHashes[mn]
+			if ok && !hasOKPodTemplateSpecHash(pod, set) {
```

Comment:
maybe worth a descriptive comment e.g. "the pod template hash doesn't correspond to the most recently deployed one--this pod is probably from an old Tilt run" or something.

Reply:
Do you think the existing comment on `hasOKPodTemplateSpecHash` suffices?
Or do you want to add a descriptive comment to each of its call sites?

Reply:
oh whoops, i missed that. ya that's fine.